### ISSUE TITLE
[Proposal] Simplified ldftn/ldvirtftn

### ISSUE THREAD

**Issue by sakno** [ltrzesniewski/InlineIL.Fody#21]
Hi Lucas!

I would like to propose simplified way to use `ldftn` and `ldvirtftn` instructions:
```csharp
public static void Ldftn<TMethod>(TMethod method) where TMethod : Delegate;
public static void Ldvirtftn<TMethod>(TMethod method) where TMethod : Delegate;
```
With these methods I can specify method directly without construction of `MethodRef`. In case of `ldvirtftn` it easily to analyze whether the specified method is virtual/abstract or not.

Alternative approach is to use factory method declared in `MethodRef` class:
```csharp
public static MethodRef Create<TMethod>(TMethod method) where TMethod : Delegate;
```

**Comment by ltrzesniewski** [comment]
Hi!

I'd prefer the factory method approach in the `MethodRef` class, since it would be more consistent with the existing API.

With the signature you suggested, you'd need to always specify type parameters explicitly, like this:

```C#
MethodRef.Create<Func<int, int>>(TargetMethod);
```

[SharpLab][1]

I suppose I could provide overloads for all `Action<...>` and `Func<...>` delegates, but it wouldn't really improve the situation:

```C#
MethodRef.Create<int, int>(TargetMethod);
```

[SharpLab][2]

Also, the IL code quickly becomes a mess if you try to use lambdas:

```C#
MethodRef.Create<int, int>(i => TargetMethod(i));
```

[SharpLab][3]

But then I guess it's the user's fault: you'll get pointers to methods of the compiler-generated class. I suppose it would be a good idea to generate an error in that case.

It would also be interesting to take a look at the new function pointers feature of C# 9, it may be helpful there:

```C#
MethodRef.Create<int, int>(&TargetMethod);
```

[SharpLab][4]

The proposal says a conversion to `void*` is possible for function pointers, but apparently the compiler doesn't allow that. I haven't really played with this feature yet, and I guess it's not complete anyway. We'll need to wait and see when it's released. Also, it doesn't look like it works for instance methods (at least in the first version that will ship).

Given that in all cases you need to specify the full function signature, do you still think this feature would be useful?


[1]: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoKwgZgAIT6BhKgbyvq8YcJXoAqMAM4YAFAEpO3DpW7z6AWRgYAFhAAmAJRgAzAHTNYAQwwwAPIQCs57ADsMaevYwA+V2IHGoAcxUBlDFNsMAkAbmkFJRV1bT1DEzNLGxcnF3dPbz8MAEk7UWM7MBhwyIVlNU0dAyMYUwtrWwc0hwyvXxUANWwoDABXY1xSuW4AXyoyukZSJGcHQSyAoIwQsRd6Ywl6AF5Xeg09Yz7cDAiRrin19uy8gqKYNfnNnb2D3SOTs/kpgDce/sGcwwCw6GG6vQGuEewOeu32h2Opyo42olCmTAqsWq7EmvBm0UqcV0LESFgEmM0HnJMU09AAtjSNFsAO6qGCwQQUjT0ED0AAiMFwMB89Re8PeiLOoyAA

[2]: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoKwgZgAIT6BhKgbyvq8YcJXoAqMAM4YAFAEpO3DpW7z6AWRgYAFhAAmAJRgAzAHTNYAQwwwAPNgB2GNPWsYAfGIHGoAcxUBlDKexgJAG5pBSUVdW09QxMzSxs7B2dXDxUASStRYyswGCCQhWU1TR0DIxhTCwcEmyS3TwwANWwoDABXY1w8uW4AXyp8ukZSJHsbQTrvXwx/MQd6Ywl6AF5Heg09Y1bcDGDurkG55Pr0zOyYWbGF5dX13U3t3flBgDdmto7RjHGUxrf23AuXyuKzWGy2OyofWolEGTEKERK7AGvGGYSKkV0LBi50IpGI9AAtuFNIsQbd7hC9jwhiN4cU9FjyrEBKRnLjiOYWasiejSTcwQ9kTS0QiGWUKpzSHYBMQ2XjJdLZYTiRo+aC7uDHtxYai6RjGRKWYrpbQ5RyjYJiCbuSq1eTNULcbSVSUDcydMJwWyAKycj1e5W867qila/Yo53o13i5lSwT+7Y+hXxkQBnkRO0CylPCMi+mYmMWC0y6UJpxiQi+4tWlOexOBjPB+2Cqk6yOigvY5MlwS0UupxMVqtxnsCPu1tO2ptZ3Y9IA

[3]: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoKwgZgAIT6BhKgbyvq8YcJXoAqMAM4YAFAEpO3DpW7z6AWRgYAFhAAmAJRgAzAHTNYAQwwwAPNgB2GNPWsYAfGOz0AvI8HGoAcxUBlDFNsMBcJCQBuaQUlFXVtPUMTM0sbOwdnVw8vXxUASStRYyswGDDI6IVlNU0dAyMYUwsHdJtM908Bbz8MADVsKAwAV2Nccqi5bgBfKkq6RlIkexscnsDg0Id6YwkO+g09YyHcDAn5ea2u3IwCopKyrZ29g90jk7PueYA3AeHR5YwqxU/UGIzGj122Reb1OVBm1Eo8yY1XidXYc14i1iNQSuhYyTKhFIxHoAFs4ppIZ5ocdYZMuEisSjanp8Y0UgJSM4icRzJzPOScVT9odaR8GZilszcWymnzSHYBMRucT5YrlWSKRphTT3hiFlKtXVZRyFYJiIraCreZz1ZaBVqdaK9fSeAbsajWQ05QIdMJadyAKx8v0BzVC57OunnSUell472mxWhk5BtWCFNOcPxJ2vMX6omGnHGxMWW3m5MiANiQjB8tKyv+1PZymRvMumPu6UlgnphuCWiN6u1vsWgdD5uCnNtmETKZAA==

[4]: https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRjgOwMYwJYD2mqADgXpjBAsgDQwh4A2APgAIBMAjALABQ/NgGYABFmRRIIziIDC/AN78RK6aLYAWEQBUUMABQBKZaqV9VFkQFkIMABYEAJgCUIYAHSyksCAB4KMLQiAQB8+gBk2lAIAOa2AMowsHjYhgDcJpYiAPTZNvZOrh5e0FT+lEGhEVGxtgCSmMhJOBDpmZa5+Q4ubp7eZQGVlGGR0XEwAGp4CPBQTG3mqgC+/O3C0lwAbMGUOmMJSfjY+gEiUIYiALwhIo5uUHBMMBmLKuunNeMNTVAtJ7vnK43O5gB5PF4WdYAN2msyYOxge1qk1hcDm/0RgOut3uj2e/BWAj463Ekgg0g41ls3SKijW6i2VIKPTAcn6fm0rmQeLCdyYEBiPgAVL5OSgeSIALbUpwXbEgsH415qDbbLqFNxs0ocrhBMXcp68iD8wVUEXaXU6LkS6XMuXA3Hg+mqpk0zUlHyiy3aDh662G/R8gXCr1631W8WGqUyxz2nGgvEQ1TrNiM9UsrWei1hvVCP2RmBGk0h7M6cPaPMRg2F6N2oHxxUvJZAA=

**Comment by sakno** [comment]
I think yes, at least for me. The main motivation from my side - this is the only way to obtain method pointer for local methods.

**Comment by sakno** [comment]
Personally I'm OK with defining delegate type. If it's too verbose you can always use alias for it.

**Comment by ltrzesniewski** [comment]
Ah, yes I see, it makes sense for local methods.

But calling local methods may be an issue. Take a look at [this example][1]: even *static* local methods are generated as *instance* methods of a compiler-generated class, as it optimizes shuffle thunks for delegate calls. [This changes][2] when you don't create a delegate from the static method.

Basically, your code would need to provide a seemingly unused parameter to the local method, it would be confusing and brittle since it would make your code depend on a compiler implementation detail. And the compiler would generate less optimal code because of the delegate on the local method.

As an aside, I agree that the first approach with the full delegate type as generic parameter is the best one, since it lets you handle methods with `ref` parameters for instance.


[1]: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoKwgZgAIT6BhKgbyvq8YcJXoAqMAM4YAFAEpO3DpW7z6AWRgYAFhAAmAJRgAzAHTNYAQwwwAPIQCs57ADsMaevYwA+V2IHGoAcxUBlDFNsMAkAbmkFJRV1bT1DEzNLGxcnF3dPbz8MAEk7UWM7MBhwyIUy+UJSJGcHQSyAoIwQsRd6Ywl6AF5Xeg09YwBXXAwIuSjajHrfFTyCophWuo7u3v7dIZGx+QBfKj3qSjpGYmi1TR1ddjLjqprlc7jdFkSLAQfYj3eYzXoAWx+Gk6AHdVDBYIIPr8QPQACIwXAwHymGCrPoDYajfZAA

[2]: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoKwgZgAIT6BhKgbyvq8YcJXoAqMAM4YAFAEpO3DpW7z6AekUBZGBgAWEACYAlGADMAdM1gBDDDAA8hAKxXsAOwxp6TjAD4PYgWagBzdQBlDAtsMAkAbmkFJVV1LT1DE3NLG3t3V3cvHz9AjABJR1EzRzAYKJiFKvlfAODQjHCxFGJKuVjBPPUikrKYFrbojtjCUiQ3Zy76jBCwsDF3ejMJegBeD3ptQzMAV1wMYc7JjGn83tD+xamV9c3tgz2Do+4AXyp36ko6RmJ6NU0On0BnYVR+YwmAMSwJYqWsAihOm8CISOnoAFtUdpVgB3DQwWCCRHaeggegAERguBg/gsMDuWx2+0OHyAA=

**Comment by ltrzesniewski** [comment]
BTW function pointers [don't have this problem][1] with local methods (static local methods are still generated as static). Maybe we should just wait for C# 9?

Also, you should be directly able to use `IL.Push` to emit a `ldftn` by [inserting an explicit cast][2]. I didn't check but I think it should just work.

[1]: https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRjgOwMYwJYD2mqADgXpjBAsgDQwh4A2APgAIBMAjALABQ/NgGYABFmRRIIziIDC/AN78RK6aLYAWEQBUUMABQBKZaqV9VFkQFkIMABYEAJgCUIYAHSyksCAB4KMLQiAQB8+gBk2lAIAOa2AMowsHjYhgDcJpYimZZsXABswZQ60XEwicnY+gEiUIYiALwhIo5uUHBMMBnmqgC+/P0CfMJimBJSMjb2Tq5gijkjeYVTDi5uct5UvtquyB0wYa1MEDE+AFTbu/vNALa2q/VNLW373RaLBdb3M+te0FvaLhBHYoa76I4nc7bIE6K6dW7fRyPZqtMDtTpvVQfZaI2Ybf5+QHAjjAuEHcEQY6nKgXIk6Emw0HwkR3aZIxool4YhbqT4rH5gfE+aHE4FCUlM8kQ6kQWkw7QM7Tixl7Zmsh4c55o14DIA==

[2]: https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRjgOwMYwJYD2mqADgXpjBAsgDQwh4A2APgAIBMAjALABQ/NgGYABFmRRIIziIDC/AN78RK6aLYAWEQBUUMABQBKZaqV9VFkQEkAMgDoACnGQALffoAmEJhADmsCAAqAB4KGFoRMIA+QwAybSgEXwgYAGUYWDxsQwBuE0sRfMs2LgA2SModROS0jPxsfTCRKEMRAF4okS8wKDgmGDzzVQBfflGBPmExTAkpGVtFIqmS8s0RJ1d9TUCRAFsUlwIPQwVx4aA===

**Comment by sakno** [comment]
Unfortunately, waiting for C# 9 is not an option. I'm using InlineIL for my library .NEXT which is based on .NET Standard 2.1. I'm not planning to upgrade to .NET 5 immediately because it's not LTS version.

**Comment by sakno** [comment]
Moreover, I don't see any evidences that function pointers will be included in C# 9. [Proposal](https://github.com/dotnet/csharplang/issues/2917) is not assigned to any milestone. Mads in [Welcome to C# 9](https://devblogs.microsoft.com/dotnet/welcome-to-c-9-0/) article doesn't mention this feature.

**Comment by ltrzesniewski** [comment]
You don't need to migrate to .NET 5 in order to use C# 9. The language version is independent of the runtime, except for a few features like default interface implementations.

Function pointers will be in C# 9 according to [the language feature status doc](https://github.com/dotnet/roslyn/blob/master/docs/Language%20Feature%20Status.md).

Anyway, I *can* implement the feature you're asking for. It could be useful to reference normal methods (and it will hopefully be future-proof), but you just need to be aware that it will lead Roslyn to generate non-optimal code for local methods, and you'll be relying on a compiler implementation detail.

In other words, it won't be great for the purpose you actually want to use it for. 😕 Function pointers would be *much* better, so if you still want to go this way, you should consider updating your code once C# 9 is released.

I'm just not entirely sure about the factory method name. I'll either add an overload to `Method`, or add a `FromDelegate` method or something. In any case, the signature should be the one you originally suggested:

```C#
public static MethodRef FromDelegate<TDelegate>(TDelegate @delegate) where TDelegate : Delegate
```

**Comment by sakno** [comment]
If referencing local methods and lambdas is too dangerous then InlineIL can analyze whether the target method is marked with `CompilerGeneratedAttribute` and produce error/warning. I can live without local methods but IMO such way to reference regular methods is less verbose.

**Comment by ltrzesniewski** [comment]
I suppose I can add a warning for these cases, mostly to prevent unintentional misuse.

You can suppress InlineIL warnings through the weaver config.

**Comment by ltrzesniewski** [comment]
I got it mostly implemented in the [methodref-delegate](https://github.com/ltrzesniewski/InlineIL.Fody/compare/methodref-delegate) branch. I just need to make sure I didn't forget anything, and maybe do some cleanup. I should release it soon.

I changed one thing though: referencing a compiler-generated method will generate an error instead of a warning. After seeing what the C# compiler generates for the various cases, I realized there's really no good way to use such a method in a sane way.

**Comment by ltrzesniewski** [comment]
I've released v1.5.0 which adds `MethodRef.FromDelegate`. Let me know if you find any issues.

**Comment by ltrzesniewski** [comment]
Oh, I forgot: there's one more thing you need to be careful about: if you try to get a reference to a virtual method override, you'll get a reference to the _base_ method instead (it's the one that Roslyn references).

This may especially be an issue for `GetHashCode` and other methods of `System.Object`, particularly for structs.

**Comment by sakno** [comment]
Thanks, Lucas!
